src/hypothesis/extra/pandas/impl.py

```diff
@@ -0,0 +1,617 @@
+    if dtype is None and not values:
+        return draw(dtype_for_elements_strategy(elements))
+    return dtype
+
+
+@check_function
+def elements_and_dtype(elements, dtype, source=None):
+
+    if source is None:
+        prefix = ''
+    else:
+        prefix = '%s.' % (source,)
+
+    if elements is not None:
+        st.check_strategy(elements, '%selements' % (prefix,))
+    else:
+        with check('dtype is not None'):
+            if dtype is None:
+                raise InvalidArgument((
+                    'At least one of %(prefix)selements or %(prefix)sdtype '
+                    'must be provided.') % {'prefix': prefix})
+
+    with check('is_categorical_dtype'):
+        if is_categorical_dtype(dtype):
+            raise InvalidArgument(
+                '%sdtype is categorical, which is currently unsupported' % (
+                    prefix,
+                ))
+
+    dtype = st.try_convert(np.dtype, dtype, 'dtype')
+
+    if elements is None:
+        elements = npst.from_dtype(dtype)
+    elif dtype is not None:
+        def convert_element(value):
+            name = 'draw(%selements)' % (prefix,)
+            try:
+                return np.array([value], dtype=dtype)[0]
+            except TypeError:
+                raise InvalidArgument(
+                    'Cannot convert %s=%r of type %s to dtype %s' % (
+                        name, value, type(value).__name__, dtype.str
+                    )
+                )
+            except ValueError:
+                raise InvalidArgument(
+                    'Cannot convert %s=%r to type %s' % (
+                        name, value, dtype.str,
+                    )
+                )
+        elements = elements.map(convert_element)
+    assert elements is not None
+
+    return elements, dtype
+
+
+class ValueIndexStrategy(st.SearchStrategy):
+    def __init__(self, elements, dtype, min_size, max_size, unique):
+        super(ValueIndexStrategy, self).__init__()
+        self.elements = elements
+        self.dtype = dtype
+        self.min_size = min_size
+        self.max_size = max_size
+        self.unique = unique
+
+    def do_draw(self, data):
+        result = []
+        seen = set()
+
+        iterator = cu.many(
+            data, min_size=self.min_size, max_size=self.max_size,
+            average_size=(self.min_size + self.max_size) / 2
+        )
+
+        while iterator.more():
+            elt = data.draw(self.elements)
+
+            if self.unique:
+                if elt in seen:
+                    iterator.reject()
+                    continue
+                seen.add(elt)
+            result.append(elt)
+
+        dtype = infer_dtype_if_necessary(
+            dtype=self.dtype, values=result, elements=self.elements,
+            draw=data.draw
+        )
+        return pandas.Index(result, dtype=dtype)
+
+
+DEFAULT_MAX_SIZE = 10
+
+
+@st.cacheable
+@st.defines_strategy
+def range_indexes(min_size=0, max_size=None):
```

Comment:
`range_indexes` is exported in the module `__all__`, and to be properly public API it also needs some documentation.